cpp/src/community/detail/mis_impl.cuh

```diff
@@ -0,0 +1,707 @@
+
+  bool debug = graph_view.local_vertex_partition_range_size() < 40;
+  if (debug) {
+    auto offsets = graph_view.local_edge_partition_view(0).offsets();
+    auto indices = graph_view.local_edge_partition_view(0).indices();
+    cudaDeviceSynchronize();
+    std::cout << "---- MIS Graph -----" << std::endl;
+    raft::print_device_vector("offsets: ", offsets.data(), offsets.size(), std::cout);
+    raft::print_device_vector("indices: ", indices.data(), indices.size(), std::cout);
+  }
+
+  rmm::device_uvector<vertex_t> remaining_vertices(local_vtx_partitoin_size, handle.get_stream());
+
+  auto vertex_begin =
+    thrust::make_counting_iterator(graph_view.local_vertex_partition_range_first());
+  auto vertex_end = thrust::make_counting_iterator(graph_view.local_vertex_partition_range_last());
+
+  // Compute out-degree and ranks
+  auto out_degrees = graph_view.compute_out_degrees(handle);
+  auto ranks       = compute_out_weight_sums<vertex_t, edge_t, weight_t, false, multi_gpu>(
+    handle, graph_view, *edge_weight_view);
+
+  // Vertices with non-zero outdegree are possible candidates for MIS.
+  remaining_vertices.resize(
+    thrust::distance(remaining_vertices.begin(),
+                     thrust::copy_if(handle.get_thrust_policy(),
+                                     vertex_begin,
+                                     vertex_end,
+                                     out_degrees.begin(),
+                                     remaining_vertices.begin(),
+                                     [] __device__(auto deg) { return deg > 0; })),
+    handle.get_stream());
+
+  if (debug) {
+    cudaDeviceSynchronize();
+    raft::print_device_vector("degrees: ", out_degrees.data(), out_degrees.size(), std::cout);
+    raft::print_device_vector(
+      "remaining_vertices: ", remaining_vertices.data(), remaining_vertices.size(), std::cout);
+  }
+
+  size_t loop_counter                              = 0;
+  vertex_t nr_remaining_vertices_in_last_iteration = 0;
+  while (true) {
+    loop_counter++;
+    // if (debug) {
+    cudaDeviceSynchronize();
+    std::cout << "------------Mis loop, counter: ---------- " << loop_counter << std::endl;
+    // }
+    //
+    // Copy ranks into temporary vector to begin with
+    //
+    // Caches for ranks
+
+    rmm::device_uvector<weight_t> temporary_ranks(local_vtx_partitoin_size, handle.get_stream());
+    thrust::copy(handle.get_thrust_policy(), ranks.begin(), ranks.end(), temporary_ranks.begin());
+
+    //
+    // Select a random set of candidate vertices
+    //
+    thrust::default_random_engine g;
+    g.seed(0);
+    thrust::shuffle(
+      handle.get_thrust_policy(), remaining_vertices.begin(), remaining_vertices.end(), g);
+
+    vertex_t nr_candidates = std::max(
+      vertex_t{1},
+      std::min(static_cast<vertex_t>((0.50 + 0.25 * loop_counter) * remaining_vertices.size()),
+               vertex_t{remaining_vertices.size()}));
```

Comment:
And just a heads-up. In multi-GPU, this logic may no longer work. While the total number of remaining vertices might still be larger than 0, the number of local remaining vertices can be 0 in some GPUs. In such a case, `nr_candidates` based on this logic (due to std::max(vertex_t{1}, ...) can overflow `remaining_vertices`.